src/Tools/dotnet-monitor/Egress/EgressValidationService.cs

```diff
@@ -0,0 +1,99 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.Diagnostics.Monitoring.EventPipe;
+using Microsoft.Diagnostics.Monitoring.WebApi;
+using Microsoft.Extensions.Hosting;
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Tools.Monitor.Egress
+{
+    internal class EgressValidationService : BackgroundService
+    {
+        private readonly IServiceProvider _serviceProvider;
+        private EgressProviderSource _egressProviderSource;
+
+        public EgressValidationService(EgressProviderSource source, IServiceProvider serviceProvider)
+        {
+            _egressProviderSource = source;
+            _serviceProvider = serviceProvider;
+        }
+
+        protected override async Task ExecuteAsync(CancellationToken token)
+        {
+            // Indicates that rules changed while validating a previous change.
+            // Used to indicate that the next iteration should not wait for
+            // another configuration change since a change was already detected.
+            bool configurationChanged = false;
+            bool initialLoad = true;
```

Comment:
@jander-msft I know we discussed this yesterday, but after playing around with initializing `EgressProviderSource` my guess is that I didn't explain my issue well enough (that or I'm doing something incorrectly here). `EgressProviderSource` is Reloaded **before** this background service starts, so when `ConfigurationChanged` is invoked there, nothing happens here. I'm currently working around this by doing a one-off initial check, but let me know if there's a more correct solution I should be using.

Reply:
Try something like this in ExecuteAsync:
- Unconditionally call Initiailize on the EgressProviderSource.
- Create a SemaphoreSlim (initial count 0, max count 1) and an event handler for the EgressProviderSource change event that will call Release on the SemaphoreSlim. Catch SemaphoreFullException since multiple reloads can overflow it.
- In a while true loop, validate the current list of egress providers and then await the SemaphoreSlim. This allows the first validation pass before waiting for the event to be raised.

There are other things to fill in like cancelling the current validation pass when the event is raised again, but that can be done with a combination of a second event handler and a linked cancellation token.

It's okay that EgressProviderSource loads before ExecutAsync starts; it shouldn't care how many times it reloaded before it does the initial validation pass. And putting a upper bound on the SemaphoreSlim ensures that the loop doesn't continuously validate the providers due to quick successive changes in configuration.

Reply:
I did something similar to what you suggested, though I didn't use a second event handler (I believe the way I'm doing it now shouldn't require it, but you can let me know if there's an issue with that approach).